lib/history/sqlite.go

```diff
@@ -0,0 +1,224 @@
+		return nil, trace.Wrap(err)
+	}
+
+	var (
+		id        int
+		timestamp time.Time
+		eventType EventType
+		node      string
+		probe     string
+		old       string
+		new       string
+	)
+
+	for rows.Next() {
+		if err := rows.Scan(&id, &timestamp, &eventType, &node, &probe, &old, &new); err != nil {
+			rows.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		var event *Event
+		switch eventType {
+		case ClusterRecovered:
+			event = NewClusterRecoveredEvent(timestamp, old, new)
+		case ClusterDegraded:
+			event = NewClusterDegradedEvent(timestamp, old, new)
+		case NodeAdded:
+			event = NewNodeAddedEvent(timestamp, node)
+		case NodeRemoved:
+			event = NewNodeRemovedEvent(timestamp, node)
+		case NodeRecovered:
+			event = NewNodeRecoveredEvent(timestamp, node, old, new)
+		case NodeDegraded:
+			event = NewNodeDegradedEvent(timestamp, node, old, new)
+		case ProbePassed:
+			event = NewProbePassedEvent(timestamp, node, probe, old, new)
+		case ProbeFailed:
+			event = NewProbeFailedEvent(timestamp, node, probe, old, new)
+		default:
+			event = NewUnknownEvent(timestamp)
+		}
+
+		events = append(events, event)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return events, nil
+}
+
+// insertEvents inserts the provided events into the timeline.
+func (t *SQLiteTimeline) insertEvents(events []*Event) error {
+
+	// prepare bulk insert statement
+	valueStrings := make([]string, 0, len(events))
+	valueArgs := make([]interface{}, 0, len(events)*eventSize)
+	for _, event := range events {
+		valueStrings = append(valueStrings, eventValueString)
+		valueArgs = append(valueArgs, event.timestamp)
+		valueArgs = append(valueArgs, event.eventType)
+		valueArgs = append(valueArgs, event.metadata["node"])
+		valueArgs = append(valueArgs, event.metadata["probe"])
+		valueArgs = append(valueArgs, event.metadata["old"])
+		valueArgs = append(valueArgs, event.metadata["new"])
+	}
+	insertEvents := fmt.Sprintf(insertIntoEvent, strings.Join(valueStrings, ","))
+
+	if _, err := t.database.Exec(insertEvents, valueArgs...); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// evictEvents deletes oldest events if the timeline is larger than its max
+// size.
+func (t *SQLiteTimeline) evictEvents() error {
+	if _, err := t.database.Exec(deleteOldFromEvent, t.size); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// createTableEvent is sql statement to create an `event` table.
+const createTableEvent = `
+CREATE TABLE event (
+	id INTEGER PRIMARY KEY,
+	timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
+	type TEXT,
+	node TEXT,
+	probe TEXT,
+	old TEXT,
+	new TEXT
+)
+`
+
+// insertIntoEvent is sql statement to insert entry into `event` table. Used for
+// batch insert statement.
+const insertIntoEvent = `
+INSERT INTO event (
+	timestamp,
+	type,
+	node,
+	probe,
+	old,
+	new
+) VALUES %s
+`
+
+// eventSize specifies the number of fields contained in an Event entry.
+const eventSize = 6
```

Comment:
Why is this needed? Pretty easy to miss this when, say, adding another field to the event.